web/src/compose_ui.js

```diff
@@ -495,6 +527,50 @@ export function format_text($textarea, type, inserted_content) {
 
             wrapSelection(field, italic_syntax);
             break;
+        case "bulleted": {
+            // Following behaviour is inspired from GitHub
+            let {before_lines, selected_lines, after_lines} = section_off_selected_lines();
+            // if there is even a single unbulleted line selected, we bullet all
+            const should_bullet = selected_lines
+                .split("\n")
+                .some((line) => !list_util.is_bulleted(line));
+            if (should_bullet) {
+                selected_lines = selected_lines
+                    .split("\n")
+                    .map((line) => "- " + line)
+                    .join("\n");
+                // add blank line between text before and list if not already present
+                if (before_lines.length && before_lines.at(-2) !== "\n") {
```

Comment:
Why is this at -2 rather than -1?

Reply:
As noted in a comment in the `section_off_selected_lines` function, "`before_lines` includes the `\n` character that separates it from `selected_lines`" and this is at -1 position. The blank line (if any) is the new line character at -2 position, which we check for here.

Do you think the code will be more intuitive if we strip the last character (`\n`) from `before_lines` returned from `section_off_selected_lines`?

Reply:
Yeah; I think that would be an improvement. The thing that made me unhappy here is that we're checking `before_lines.length`, which just confirms it's at least one character, and then accessing it in a way that requires that it be at least two characters, which is the kind of thing that really looks like a bug (even if it might not be in practice if `before_lines` is never exactly 1 character)

Reply:
Right, it does read as very confusing, thanks for pointing this out!

This needed a fair bit of careful refactoring, but I added to the tests which all pass, and have tested manually as thoroughly as I could too